### PR TITLE
feat(daemon): E-lite root-cause telemetry for inactivity_timeout attribution

### PR DESCRIPTION
**Pull request by lefarcen**
## Why

**Author use case:** I've been doing reliability root-cause analysis off the daily stability report. On the latest release (0.15.0) `inactivity_timeout` is the one top fault we still can't attribute — 275 of its failures die in the `tool_execution` phase, but `last_observed_phase` only tells us *which* phase stalled, not *why*. I literally cannot tell an approval-hang from a provider-stall from a tool-result-that-never-came-back.

**Pain:** This blind spot has persisted for a long time. Every reliability round ends with "inactivity_timeout in tool_execution — root cause unknown, needs finer telemetry." This PR adds exactly that telemetry so the *next* round can attribute it and we can ship a real fix instead of guessing.

## What users will see

Nothing user-facing. This is internal analytics instrumentation on the `run_finished` event. The user-visible payoff is downstream: once this ships and adopts, the reliability dashboard can split `inactivity_timeout` into distinct root causes and we can fix the dominant one.

## What it adds

Four fields on `run_finished` that separate WHY a stalled run died (the phase already tells us WHICH phase):

| field | meaning | how populated |
|---|---|---|
| `approval_requested` | an ACP approval/permission gate fired | ACP session emits an `acp_approval_request` diagnostic; folded in by the diagnostics summarizer. Stream/CLI runtimes bypass gates, so `false` there = "not observed" |
| `stdin_backpressure` | prompt write to child stdin was queued (pipe buffer full) | `child.stdin.write` return / `writableNeedDrain`, stored on the run |
| `tool_result_sent` | a tool_result came back for the run's **last** tool_use | derived from `run.events` (a stall with `tool_call_seen && !tool_result_sent` = tool-result-not-delivered) |
| `last_progress_age_ms` | age of last agent activity at finish | the inactivity-watchdog clock, read at finish |

## Surface area

- [x] Analytics contract (`packages/contracts` — `RunFinishedProps`)
- [x] Daemon (`apps/daemon` — instrumentation + summarizer)
- [ ] Web UI — n/a
- [ ] `od` CLI — n/a. This is a telemetry field, not a user-invokable capability, so the UI/CLI dual-track rule does not apply (no new endpoint, no user surface).
- [ ] i18n / env vars / extension points — none

## Testing

- `pnpm --filter @open-design/contracts typecheck` — exit 0
- `pnpm --filter @open-design/daemon typecheck` — exit 0
- `apps/daemon/tests/run-diagnostics.test.ts` — 9 passed (new test covers `tool_result_sent` true/false + `approval_requested`)
- `packages/contracts/tests/analytics-run-finished-contract.test.ts` — 4 passed
- `pnpm guard` — 86 pass, 0 fail

Runtime end-to-end (does `approval_requested` actually fire on a real ACP approval, do the fields flow to PostHog) verifies once this ships in a release with traffic; the ACP diagnostic mirrors the existing `acp_raw_event_shape` / `acp_artifact_text_suppression` diagnostics that already reach `run.events`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by lefarcen** [comment]
Thanks — the root-cause breakdown in the `Why` section is clear and the validation notes are already helpful.

One small PR-body follow-up before pool review picks this up: could you make the `Surface area` checklist explicit so the parser sees the touched areas clearly? Right now `Validation` is covered, but the checklist still looks missing from the template's point of view.

File: apps/daemon/src/run-diagnostics.ts [hunk 1]
```diff
@@ -183,7 +194,16 @@ export function summarizeRunDiagnosticsForAnalytics(args: {
       const delta = typeof data.delta === 'string' ? data.delta : '';
       if (delta.length > 0) userVisibleOutputSeen = true;
     }
-    if (data.type === 'tool_use') toolCallSeen = true;
+    if (data.type === 'tool_use') {
+      toolCallSeen = true;
+      toolResultAfterLastToolUse = false;
+    }
+    if (data.type === 'tool_result' && toolCallSeen) {
+      toolResultAfterLastToolUse = true;
```
Comment (nettee) [inline apps/daemon/src/run-diagnostics.ts]: `tool_result_sent` is documented here as "a tool_result followed the run's LAST tool_use", but this implementation flips it to `true` for any `tool_result` after we've seen at least one tool call. In a multi-tool turn like `tool_use(A)`, `tool_use(B)`, `tool_result(A)`, this marks the metric as delivered even though the last tool `B` is still outstanding. That breaks the exact attribution this PR is trying to add: multi-tool stalls will be bucketed as "provider stalled after delivery" instead of "tool result never came back". The normalized events already carry `toolUseId`, so the fix should track the most recent tool-use id and only set `tool_result_sent` when the matching result arrives; please add a regression test for the two-tool case as well.

<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

**Review by nettee — CHANGES_REQUESTED** [review]
I found one blocking correctness issue in the new telemetry derivation; details are inline.

<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

File: apps/daemon/src/run-diagnostics.ts (hunk 1 above)
Comment (lefarcen) [reply apps/daemon/src/run-diagnostics.ts]: Good catch — confirmed and fixed in c80c853. `tool_result_sent` is now paired by id (`tool_use.id` ↔ `tool_result.toolUseId`, the same pairing `summarizeRunTimingAnalytics` uses): it's true only when **every** committed tool_use received a matching tool_result (outstanding set empty). Your exact case — `tool_use(A), tool_use(B), tool_result(A)` — now correctly stays `false` while B is outstanding, and there's a regression test for it plus parallel-all-resolved and sequential-last-outstanding.

File: apps/daemon/src/run-diagnostics.ts [hunk 2]
```diff
@@ -183,7 +199,17 @@ export function summarizeRunDiagnosticsForAnalytics(args: {
       const delta = typeof data.delta === 'string' ? data.delta : '';
       if (delta.length > 0) userVisibleOutputSeen = true;
     }
-    if (data.type === 'tool_use') toolCallSeen = true;
+    if (data.type === 'tool_use') {
+      toolCallSeen = true;
+      sawAnyToolUse = true;
+      if (typeof data.id === 'string') outstandingToolUses.add(data.id);
```
Comment (nettee) [inline apps/daemon/src/run-diagnostics.ts]: This still misclassifies stalled tool runs when the runtime emits a `tool_use` without a string id. In this hunk we set `sawAnyToolUse = true` for every `tool_use`, but we only track outstanding work when `typeof data.id === 'string'`, so `tool_result_sent` falls through to `true` whenever an unpaired/null-id tool call never gets a result. That breaks the new attribution goal by bucketing those stalls as "tool result delivered" instead of "tool result never came back". The daemon already has adapters that emit nullable ids on degraded provider events, for example `apps/daemon/src/agent-protocol/pi-rpc/events.ts` (`raw.toolCallId ?? null`) and `apps/daemon/src/copilot-stream.ts` (`data.toolCallId ?? null`), so this case is reachable today. Please treat a `tool_use` without a usable id as unresolved for this metric (for example, gate `sawAnyToolUse` on the same string-id check or add an explicit `missingToolUseId` flag that forces `tool_result_sent` to `false`) and add a regression test for the null-id path.

<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

**Review by nettee — CHANGES_REQUESTED** [review]
I found one remaining blocking telemetry-correctness issue in the updated `tool_result_sent` derivation; details are inline.

<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

File: apps/daemon/src/server.ts [hunk 3]
```diff
@@ -7973,7 +7980,10 @@ export async function startServer({
         }
         run.stdinOpen = true;
       } else {
+        // `end()` returns the stream (not a boolean), so read backpressure from
+        // `writableNeedDrain` right after queueing the final chunk (see above).
         child.stdin.end(composed, 'utf8', markStdinWriteEnd);
+        run.stdinBackpressure = child.stdin.writableNeedDrain === true;
```
Comment (nettee) [inline apps/daemon/src/server.ts]: This branch never records `stdin_backpressure` for the default text-input runtimes. `child.stdin.end(composed, ...)` does not surface the boolean backpressure signal that `write(...)` returns, and `writableNeedDrain` is already `false` again immediately after `end(chunk)` even when the same payload would have made `write(chunk)` return `false`. That means the new field stays `false` for non-`stream-json` runs, so the reliability dashboard will miss exactly the `stdin_write` stalls this PR is trying to attribute. Please split this into an explicit `write` + `end` sequence (or another path that captures the boolean return from the final write) and add a regression test around the non-`stream-json` branch so the telemetry is correct for the common runtime path.

<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

**Review by nettee — CHANGES_REQUESTED** [review]
I found one blocking correctness issue in the new `stdin_backpressure` telemetry; details are inline.

<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

File: apps/daemon/src/server.ts (hunk 3 above)
Comment (lefarcen) [reply apps/daemon/src/server.ts]: Right again — fixed in 0ab14eb. I'd checked the null-id path and wrongly judged it acceptable; your two references settle it: `agent-protocol/pi-rpc/events.ts` and `copilot-stream.ts` both emit `toolCallId ?? null` for **tool_use.id and tool_result.toolUseId symmetrically**, so id-less events are a real degraded-provider path, not a hypothetical.

Id-less tool events are now paired by count alongside the id-keyed set:

```
tool_result_sent = sawAnyToolUse
  && outstandingToolUseIds.size === 0      // id-paired work all resolved
  && idlessToolResults >= idlessToolUses   // id-less work paired by count
```

An unpaired id-less tool call now correctly stays `false` (it can no longer fall through to "delivered" and mask the stall), while an id-less pair that did resolve still reports `true`. Both cases have regression tests, alongside the parallel/multi-tool ones from your first round.

**Comment by lefarcen** [comment]
Re-triggering review: the two blocking findings from the earlier rounds are both fixed and pushed in `0ab14eb` (id-based tool pairing, then count-pairing for degraded events that emit null ids on both sides). Full verification on that HEAD: daemon + contracts typecheck clean, `run-diagnostics` 9 tests pass (including regression cases for both findings), `pnpm guard` 86/0.

Context on urgency: this telemetry has now missed both 0.15.1 and 0.16.0 while sitting on a stale CHANGES_REQUESTED, and it can only start producing data once it ships in a release. Next window is the Thursday 0.16.1 patch cut from main.

File: apps/daemon/src/server.ts (hunk 3 above)
Comment (lefarcen) [reply apps/daemon/src/server.ts]: Correct, and this one was the worst of the three — the field was dead on the path that matters. Fixed in the commit above.

`end(chunk)` gave back the stream, and `writableNeedDrain` had already reset by the time it returned, so non-`stream-json` runs always reported `false`. Since Claude is the only runtime on the stream-json path, `stdin_backpressure` was blind on essentially every run whose `stdin_write` stall it was added to attribute.

The write and close are now separate, through `writePromptAndEndStdin` in `chat-run-lifecycle.ts`, returning the boolean the pipe actually gave us. I extracted it rather than inlining it so it is unit-testable, per your request for a regression test on that branch — three cases: backpressured write, accepted write, and the flush callback still firing so the `stdin_write_end` lifecycle mark is unaffected. Verified red against the old behavior (`expected false to be true`) and green after.

Also worth saying plainly: three rounds, three real defects, all mine. Thanks for holding the line on this one.

**Review by nettee — APPROVED** [review]
@lefarcen I re-reviewed the final head with the stdin backpressure fix in place. The default text-input path now captures the real `stdin.write()` return value via `writePromptAndEndStdin`, keeps the `stdin_write_end` lifecycle callback intact, and the surrounding daemon/contracts coverage now exercises the earlier multi-tool, id-less, and stdin regression cases that came up across the prior review rounds.

I wasn't able to run the local `pnpm` test/typecheck commands in this prepared reviewer worktree because `node_modules` are absent here (`vitest`/`esbuild` are not installed), but the live diff ranges are internally consistent on `9004cc3eb8bbf0d31eaa0e5c9b1465981121a399`. Nice recovery on a tricky telemetry change.

<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

**Comment by lefarcen** [comment]
🧪 Queued for QA validation — this PR has changes that need a manual QA pass before it's merged. Nothing needed from you; we'll update here once it's validated. Thanks for the contribution! 🙏

**Comment by AmyShang-alt** [comment]
@lefarcen

Thanks for the contribution. I completed QA validation for this PR.

QA Acceptance Record

Scope:
- Validated the daemon/contracts telemetry changes for `run_finished` inactivity-timeout attribution on head `9004cc3eb8bbf0d31eaa0e5c9b1465981121a399`.

Verified:
- `pnpm --filter @open-design/contracts typecheck` passed.
- `pnpm --filter @open-design/daemon typecheck` passed.
- `pnpm --filter @open-design/daemon exec vitest run tests/run-diagnostics.test.ts tests/chat-run-lifecycle-stdin.test.ts` passed: 2 files, 12 tests.
- `pnpm --filter @open-design/contracts exec vitest run tests/analytics-run-finished-contract.test.ts` passed: 1 file, 4 tests.
- `pnpm guard` passed: 86 pass, 0 fail.
- PostHog read access was checked: recent production `run_finished` events are queryable, and the new fields are not present in production yet, which is expected before this PR ships.

Not verified:
- End-to-end production ingestion of the four new fields after release; this requires the PR to ship and real traffic to emit the updated `run_finished` payload.

Risks / notes:
- The previous review findings around multi-tool pairing, id-less tool events, and default stdin backpressure are covered by regression tests.
- The remaining risk is post-release telemetry adoption, not merge-blocking for this PR.

Conclusion:
- Accepted / QA passed.

**Review by AmyShang-alt — APPROVED** [review]
